### PR TITLE
create /report command

### PR DESCRIPTION
**Pull request by SpEcHiDe**

### CODE REVIEW

**Comment by Lonami** [comment]
This needs improvements:
* A cooldown (at least 10 minutes).
* If the message being replied was in the N (probaby 10) most recent reports, don't report it again.
* For the time being, only allow reporting messages with media, usernames or links. This will prevent abuse of the report command on normal messages.

**Comment by nitanmarcel** [comment]
> This needs improvements:
> 
> * A cooldown (at least 10 minutes).
> * If the message being replied was in the N (probaby 10) most recent reports, don't report it again.
> * For the time being, only allow reporting messages with media, usernames or links. This will prevent abuse of the report command on normal messages.

What about instead of doing a cooldown and counting the x number of reported message the bot will save the last reported message and not allowing it to be reported again?

**Comment by nitanmarcel** [comment]
Also the bot should PM admins instead of mentioning them (just my thoughts)

**Comment by nitanmarcel** [comment]
And one more thing @SpEcHiDe, Silvia commands always starts with hashtag and having something different might create confusion

File: report.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+async def init(bot):
+    # utils = modules['utils']
+
+    @bot.on(events.NewMessage(pattern='\/report'))
+    async def handler(event):
+        # mentions = '<a href="tg://user?id={}">{}</a> '.format(
+        #     event.sender_id,
+        #     utils.get_display(event.sender)
+        # )
+        if not event.reply_to_msg_id:
```
Comment (Lonami) [inline report.py:13]: `get_reply_message` will return `None` if there is no reply. This check can be removed and instead the `None` check added below.

File: report.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+            return
+        reply_message = await event.get_reply_message()
+        if not (
+            reply_message.media or 
+            reply_message.entities
```
Comment (Lonami) [inline report.py:18]: This allows someone to report messages with bold, italics, etc. text, which is not desirable.

**Review by Lonami — CHANGES_REQUESTED** [review]
Changes still lack a 10-minute cooldown.

**Comment by Lonami** [comment]
> What about instead of doing a cooldown and counting the x number of reported message the bot will save the last reported message and not allowing it to be reported again?

What you suggest is exactly my second point, but rather than last message, a bigger amount of recent reports.

File: report.py [hunk 3]
```diff
@@ -1,21 +1,40 @@
...
         #     event.sender_id,
         #     utils.get_display(event.sender)
         # )
-        if not event.reply_to_msg_id:
-            return
         reply_message = await event.get_reply_message()
+        if not reply_message:
+            return
+        is_link = any([
+            True if isinstance(
```
Comment (Lonami) [inline report.py:24]: `True if condition() else False` is the same as just `condition()`. Please change this.

File: report.py [hunk 4]
```diff
@@ -1,21 +1,40 @@
...
+                e, 
+                (
+                    MessageEntityEmail,
+                    MessageEntityMention,
+                    MessageEntityMentionName,
+                    MessageEntityTextUrl,
+                    MessageEntityUrl
+                )
+            ) else False for e in reply_message.entities or []
```
Comment (Lonami) [inline report.py:33]:
```python
for e in (reply_message.entities or ())
```

File: report.py [hunk 5]
```diff
@@ -1,21 +1,40 @@
...
         # mentions = '<a href="tg://user?id={}">{}</a> '.format(
         #     event.sender_id,
         #     utils.get_display(event.sender)
         # )
-        if not event.reply_to_msg_id:
-            return
         reply_message = await event.get_reply_message()
+        if not reply_message:
+            return
+        is_link = any([
```
Comment (Lonami) [inline report.py:23]: It is not necessary to create a list here (so `any([…])` becomes just `any(…)`).

**Review by Lonami — CHANGES_REQUESTED** [review]

File: report.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    MessageEntityMentionName,
+    MessageEntityTextUrl,
+    MessageEntityUrl
+)
+
+
+async def init(bot):
+    # utils = modules['utils']
+    
+    reporting_time = {}  # user id to time they asked
```
Comment (Lonami) [inline report.py:15]:
```suggestion
    reporting_time = {}  # chat_id to time of report
```

File: report.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    reporting_delay = 10 * 60
```
Comment (Lonami) [inline report.py:16]:
```suggestion
    REPORT_DELAY = 10 * 60
```

File: report.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+    @bot.on(events.NewMessage(pattern='(?\/|#)report'))
+    async def handler(event):
+        # mentions = '<a href="tg://user?id={}">{}</a> '.format(
+        #     event.sender_id,
+        #     utils.get_display(event.sender)
+        # )
+        delta = time.time() - reporting_time.get(event.sender_id, 0)
+        if delta < reporting_delay:
```
Comment (Lonami) [inline report.py:25]:
```suggestion
        if delta < REPORT_DELAY:
```

File: report.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    reporting_time = {}  # user id to time they asked
+    reporting_delay = 10 * 60
+
+    @bot.on(events.NewMessage(pattern='(?\/|#)report'))
+    async def handler(event):
+        # mentions = '<a href="tg://user?id={}">{}</a> '.format(
+        #     event.sender_id,
+        #     utils.get_display(event.sender)
+        # )
+        delta = time.time() - reporting_time.get(event.sender_id, 0)
```
Comment (Lonami) [inline report.py:24]:
```suggestion
        delta = time.time() - reporting_time.get(event.chat_id, 0)
```

File: report.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+                    MessageEntityMentionName,
+                    MessageEntityTextUrl,
+                    MessageEntityUrl
+                )
+            ) for entity in (reply_message.entities or [])
+        )
+        if not (
+            reply_message.media or 
+            is_link
+        ):
```
Comment (Lonami) [inline report.py:46]:
```suggestion
        has_links = any(isinstance(entity, (
            MessageEntityEmail,
            MessageEntityMention,
            MessageEntityMentionName,
            MessageEntityTextUrl,
            MessageEntityUrl,
        )) for entity in (reply_message.entities or ()))
        if not reply_message.media and not has_links:
```

File: report.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+            return
+        mentions = 'Reported to admins.'
+        async for x in event.client.iter_participants(event.chat_id, filter=ChannelParticipantsAdmins):
+            if not x.bot:
+                mentions += f"[\u2063](tg://user?id={x.id})"
+        await reply_message.reply(mentions)
+        reporting_time[event.sender_id] = time.time()
```
Comment (Lonami) [inline report.py:53]:
```suggestion
        reporting_time[event.chat_id] = time.time()
```

**Review by Lonami — CHANGES_REQUESTED** [review]

**Comment by Lonami** [comment]
Still missing the ability to ignore the most recent messages that have already been reported. We likely want a bounded `collections.deque` here.

**Pull request closed without merging**